Reimport: keep false positive and out of scope history #3848

macedogm opened the pull request:
Fix issue https://github.com/DefectDojo/django-DefectDojo/issues/3848 regarding the reimport of reports where previous findings marked as false positive and/or out of scope are re-activated instead of ignored. Now the history of false positive and out of scope findings is properly preserved.

valentijnscholten commented:
Thanks. Please note that almost the same code is also located in `test/views.py`. Could you also add a few lines to the description of what problem is being solved?

macedogm commented:
Thanks @valentijnscholten . I'll review your comment. Right now I'm trying to fix the unit test, it's failing when it's executed simulating the UI (`FAIL: test_import_reimport_keep_false_positive_and_out_of_scope (dojo.unittests.test_import_reimport.ImportReimportTestUI)`). I'm trying to understand the logic behind this, since in the UI there is no way to re-import a report to the same `test_id`. Every import in the UI will generate a new test, with a new `test_id`. Right? Or am I missing something?

valentijnscholten commented:
You can reimport into the same test. Look for "re-upload" as it's called that in some places.

macedogm commented:
@valentijnscholten Thanks for pointing me to the right direction. It took me a while to follow through the code and understand the flow that is happening in `ImportReimportTestUI`, `test/views.py` and why the test was failing. Please let me know what do you think now.

macedogm commented:
ℹ️  Might make sense to merge https://github.com/DefectDojo/django-DefectDojo/pull/3854 before and then I update this MR accordingly.

valentijnscholten commented:
Thanks, this is a good improvement. Weird how nobody ever ran into this, might be that not many people are using reimport? We are improving reimport big time as I believe it is very powerful and useful.
Looking again at this I realize `risk_accepted`  should be added to this same if clause, agree?
For `duplicate` I am in doubt, might be that it needs to reactivated when needed.
And ideally we also test the opposite of the if, so that if a finding is not false_p/out_of_scope/risk_accepted, reactivation will be performed? In a way I think that is already tests by the pre existing test cases.

macedogm commented:
> We are improving reimport big time as I believe it is very powerful and useful.

For our team the reimport feature is super useful and important. We ran dozens of daily scans and we only noticed the issue when we marked some findings as false positives and they started to appear again. So I took the time to investigate and tried to fix this.
We were also hit by an issue related to the comment below during a reimport of a GitLab dependency scanning report (I submitted this parser a few weeks ago): https://github.com/DefectDojo/django-DefectDojo/blob/33fc63962b8e901bdf2b31bad735277e48951b1c/dojo/api_v2/serializers.py#L1420-L1432 . I also plan to submit a PR with a proposal to improve this during the week.


> Looking again at this I realize risk_accepted should be added to this same if clause, agree?

Yes, I agree. Missed that one. Will update the PR.

> For duplicate I am in doubt, might be that it needs to reactivated when needed.

In theory a `duplicate` can be reactivated and the `dedup` algorithm will take care of deduplicating it later. Right?

> And ideally we also test the opposite of the if, so that if a finding is not false_p/out_of_scope/risk_accepted, reactivation will be performed? In a way I think that is already tests by the pre existing test cases.

That's a good point. I can add a test for it inside test case. What do you think?

Please let me know if you agree with the above comments and I'll work on those improvements tomorrow or Tuesday.

valentijnscholten commented:
> We were also hit by an issue related to the comment below during a reimport of a GitLab dependency scanning report (I submitted this parser a few weeks ago):
> 
> https://github.com/DefectDojo/django-DefectDojo/blob/33fc63962b8e901bdf2b31bad735277e48951b1c/dojo/api_v2/serializers.py#L1420-L1432


I would wait maybe for #3753 as that will change the matching logic for reimport.

macedogm commented:
> Looking again at this I realize risk_accepted should be added to this same if clause, agree?

Added validation to skip the reimport of findings that are `false_p` or `out_of_scope` or `risk_accepted`.

> And ideally we also test the opposite of the if, so that if a finding is not false_p/out_of_scope/risk_accepted, reactivation will be performed? In a way I think that is already tests by the pre existing test cases.

Added inside the test an assertion to make sure that one finding that was deleted was properly reimported. Do you agree with this (please let me know if the test got too big)?
I had to add some support functions in `unittests/dojo_test_case.py` in order to be able to:
1. Delete a finding.
2. Get the `engagement_id` from a `test_id` through API request.
3. Get the `product_id` from an `engagement_id` also through API request.
4. Update a product configuration through the `product_id` to allow findings to be marked as `risk_accepted` through `enable_simple_risk_acceptance` flag.

> For duplicate I am in doubt, might be that it needs to reactivated when needed.

I'm also in doubt about this, because if the finding was marked as `duplicate`, then its duplicated finding will:
1. Be reopened through the reimport, in case it was deleted; or
2. Be kept inactive, in case it's a `false_p` or `out_of_scope` or `risk_accepted`

So I feel that this behaviour is already tested. What to you think?

Gby56 commented:
This seems to have bugs with some parsers, the Veracode parser definitely has the problem because "Mitigated by Design" doesn't map to false positive, out of scope, or risk_accepted, so they always get reopened...
Also, if we're completely skipping a finding for these justifications, we might be missing some updated justifications on a reimport ? I would expect a date comparison to see if the scanner data has changed in the meantime
